ground/src/test/java/com/google/android/ground/ui/datacollection/DataCollectionViewModelTest.kt

```diff
@@ -0,0 +1,123 @@
+import com.google.android.ground.model.submission.Submission
+import com.google.android.ground.model.task.Task
+import com.google.android.ground.persistence.local.LocalDataStore
+import com.google.android.ground.persistence.local.LocalDataStoreModule
+import com.google.android.ground.repository.SubmissionRepository
+import com.google.common.truth.Truth.assertThat
+import dagger.hilt.android.testing.BindValue
+import dagger.hilt.android.testing.HiltAndroidTest
+import dagger.hilt.android.testing.UninstallModules
+import io.reactivex.Single
+import java8.util.Optional
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.whenever
+import org.robolectric.RobolectricTestRunner
+import java.util.*
+import javax.inject.Inject
+
+@HiltAndroidTest
+@UninstallModules(LocalDataStoreModule::class)
+@RunWith(RobolectricTestRunner::class)
+class DataCollectionViewModelTest : BaseHiltTest() {
+
+    private val surveyId: String = "123"
+    private val loiId: String = "456"
+    private val submissionId: String = "789"
+    private val jobName = "jobName"
+    private val loiName = "loiName"
+    private val args = DataCollectionFragmentArgs.Builder(surveyId, loiId, submissionId).build()
+    private val auditInfo = TestModelBuilders.newAuditInfo()
+        .setUser(TestModelBuilders.newUser().setId("user1").build())
+        .setClientTimestamp(Date(100))
+        .setServerTimestamp(Optional.of(Date(101)))
+        .build()
+    private val survey = Survey.newBuilder().setId(surveyId).setTitle("surveyTitle")
+        .setDescription("surveyDescription").build()
+    private val submission = Submission.newBuilder()
+        .setId(submissionId)
+        .setSurvey(survey)
+        .setLocationOfInterest(
+            PointOfInterest.newBuilder().setCaption(loiName).setSurvey(survey).setCreated(auditInfo)
+                .setLastModified(auditInfo)
+                .setPoint(Point.newBuilder().setLatitude(0.0).setLongitude(0.0).build())
+                .setJob(Job.newBuilder().setName(jobName).setId("jobId").build())
+                .setId(loiId)
+                .build()
+        )
+        .setCreated(auditInfo)
+        .setLastModified(auditInfo)
+        .setTask(Task.newBuilder().setId("taskId").build())
+        .build()
+
+    @BindValue
+    @Mock
+    lateinit var localDataStore: LocalDataStore
+
+    @BindValue
+    @Mock
+    lateinit var submissionRepository: SubmissionRepository
+
+    @Inject
+    lateinit var dataCollectionViewModel: DataCollectionViewModel
+
+    @Before
+    override fun setUp() {
+        super.setUp()
+
+        whenever(
+            submissionRepository.createSubmission(
+                any(),
+                any(),
+                any()
+            )
+        ).doReturn(Single.create {
+            submission
+        })
+    }
+
+    @Test
```

Comment:
For some reason both of these tests are always failing because the actual jobName and loiName are both null.

I wonder if I might be missing something in terms of waiting for the result of the load to propagate? Let me know if you have any ideas!

Reply:
Actually, that's probably expected, since `observe()` always replays the latest value. If there's a race condition with `loadSubmissionDetails()` then the second call will get the initial value as well. Since this is async logic, we probably need something that captures value up to some timeout.

Reply:
Looks like you need to add and remove and `Observer`, and to use `InstantTaskExecutorRule` to ensure all background tasks are run on the same blocking thread:
https://developer.android.com/codelabs/advanced-android-kotlin-training-testing-basics#8

Reply:
So I tried following that codelab, but the test fails with the message "LiveData value was never set." after I use the LiveDataTestUtil class they provide and the `InstantTaskExecutorRule`.

I also tried the step before the LiveDataTestUtil which was basically the same as my previous test, but with the `InstantTaskExecutorRule` and the values were still null.

Reply:
@dturner Any idea what we might be missing here?

Reply:
> So I tried following that codelab, but the test fails...

Is it the codelab which is failing, or just when you copy/paste code from the codelab into the ground codebase?

Reply:
The tests in DataCollectionViewModelTest are failing when I am trying to use the copy pasted LiveDataTestUtil code from the codelab.

The jobName and loiName LiveData are apparently never getting set (at least according to the LiveDataTestUtil).

Reply:
Few thoughts off the top of my head: 

1) Is 

```
@get:Rule
var instantTaskExecutorRule = InstantTaskExecutorRule()
```

the first line in `DataCollectionViewModelTest`?

2) From the codelab: 
 
> This rule runs all Architecture Components-related background jobs in the same thread 

On what thread are you writing the initial livedata values? It could be that `InstantTaskExecutorRule` is not forcing that thread to run synchronously. 

3) Are you using the androidx version of [`InstantTaskExecutorRule`](https://developer.android.com/reference/androidx/arch/core/executor/testing/InstantTaskExecutorRule) - the one in the codelab is deprecated?